timeoutreader_windows.go

```diff
@@ -0,0 +1,259 @@
+	if timeout < 0 {
+		panic("Timeout must be greater than or equal to 0")
+	}
+	// if we don't set console mode, the inputrecord struct's values become.... corrupted?
+	err := windows.SetConsoleMode(windows.Handle(stream.Fd()), uint32(fdwMode))
+	if err != nil {
+		// TODO: decide how to handle this
+	}
+	return &TimeoutReaderWindows{
+		handle:              syscall.Handle(os.Stdin.Fd()),
+		timeoutMilliseconds: safecast.MustConv[uint32](timeout.Milliseconds()),
+		blocking:            timeout == 0,
+		ostream:             stream,
+		resizeChannel:       make(chan WindowsResize, 1),
+	}
+}
+
+func (tr *TimeoutReaderWindows) IsClosed() bool {
+	return tr.ostream == nil
+}
+
+func (tr *TimeoutReaderWindows) Close() error {
+	var err error
+	if tr.blocking && tr.ostream != nil {
+		err = tr.ostream.Close()
+		tr.ostream = nil
+	}
+	return err
+}
+
+func (tr *TimeoutReaderWindows) RawMode() error {
+	return nil
+}
+
+func (tr *TimeoutReaderWindows) NormalMode() error {
+	return nil
+}
+
+func (tr *TimeoutReaderWindows) StartDirect() {
+}
+
+func (tr *TimeoutReaderWindows) ChangeTimeout(timeout time.Duration) {
+	tr.timeoutMilliseconds = safecast.MustConv[uint32](timeout.Milliseconds())
+}
+
+func (tr *TimeoutReaderWindows) ReadBlocking(p []byte) (int, error) {
+	// TODO: figure out why we need these two lines every time we ReadBlocking but not for ReadWithTimeout
+	fdwMode := windows.ENABLE_EXTENDED_FLAGS
+	_ = windows.SetConsoleMode(windows.Handle(tr.handle), uint32(fdwMode))
+	var iR InputRecord
+	var read uint32
+	err := ReadConsoleInput(tr.handle, &iR, 1, &read)
+	if err != nil {
+		log.Errf("ReadConsoleInput error: %v", err)
+		return 0, err
+	}
+	nilCheck := InputRecord{}
+	if iR == nilCheck {
+		return 0, nil // timeout case
+	}
+	return iR.Read(p, tr.resizeChannel)
+}
+
+func (tr *TimeoutReaderWindows) PrimeReadImmediate(buf []byte) {
+	tr.buf = buf
+}
+
+func (tr *TimeoutReaderWindows) Read(buf []byte) (int, error) {
+	tr.mut.Lock()
+	defer tr.mut.Unlock()
+	if tr.blocking {
+		return tr.ReadBlocking(buf)
+	}
+	return ReadWithTimeout(tr.handle, tr.timeoutMilliseconds, buf, tr.resizeChannel)
+}
+
+func (tr *TimeoutReaderWindows) ReadImmediate() (int, error) {
+	if tr.blocking {
+		return tr.ReadBlocking(tr.buf)
+	}
+	return ReadWithTimeout(tr.handle, 0, tr.buf, tr.resizeChannel)
+}
+
+func (tr *TimeoutReaderWindows) ReadWithTimeout(buf []byte) (int, error) {
+	return ReadWithTimeout(tr.handle, tr.timeoutMilliseconds, buf, tr.resizeChannel)
+}
+
+const (
+	WAITOBJECT0 = 0x00000000
+	WAITTIMEOUT = 0x00000102
+)
+
+func ReadWithTimeout(handle syscall.Handle, ms uint32, buf []byte, resizeChannel chan WindowsResize) (int, error) {
+	var iR InputRecord
+	var read uint32
+	event, err := windows.WaitForSingleObject(windows.Handle(handle), ms)
+	if event == WAITTIMEOUT {
+		return 0, nil
+	}
+	if err != nil {
+		return 0, err
+	}
+	err = ReadConsoleInput(handle, &iR, 1, &read)
```

Comment:
reading only 1 byte/key at a time will probably suck perf wise when for instance copy pasting?